README.md
# bip

[![npm version](https://img.shields.io/npm/v/bip.svg?style=flat-square)](https://www.npmjs.com/package/bip)
[![npm downloads](https://img.shields.io/npm/dm/bip.svg?style=flat-square)](https://www.npmjs.com/package/bip)
[![Build Status](https://img.shields.io/travis/lgaticaq/bip.svg?style=flat-square)](https://travis-ci.org/lgaticaq/bip)
[![Coverage Status](https://img.shields.io/coveralls/lgaticaq/bip/master.svg?style=flat-square)](https://coveralls.io/github/lgaticaq/bip?branch=master)
[![Code Climate](https://img.shields.io/codeclimate/github/lgaticaq/bip.svg?style=flat-square)](https://codeclimate.com/github/lgaticaq/bip)
[![devDependency Status](https://img.shields.io/david/dev/lgaticaq/bip.svg?style=flat-square)](https://david-dm.org/lgaticaq/bip#info=devDependencies)

> Get balance of bip card (Chile)

## Installation

```bash
npm i -S bip
```

## Use

[Try on RunKit](https://runkit.com/npm/bip)
```js
const bip = require('bip');

const number = 11111111;

bip(number)
  .then(console.log)
  .catch(console.error);
```

Result:
```js
{
  number: XXXXX, // a number
  balance: XXXXX, // a number
  date: XXXXX, // a date
  message: XXXXX, // a string
  valid: XXXXX // a boolean
}
```

## Licencia

[MIT](https://tldrlegal.com/license/mit-license)


CHANGELOG.md
#### 3.0.2 (2018-09-21)

##### Chores

* **deps:**
  *  update dependency lint-staged to v7.3.0 ([d54bc2eb](https://github.com/lgaticaq/bip/commit/d54bc2eb6124f4dcca0d165094971b03c38b59ed))
  *  update dependency eslint to v5.6.0 ([13dbb772](https://github.com/lgaticaq/bip/commit/13dbb772ccd6e1c6c68787358a922ef355c645a6))
  *  update dependency eslint to v5 ([99dba2f4](https://github.com/lgaticaq/bip/commit/99dba2f48168fce355b063f9f3ee13318256058f))
  *  update dependency eslint-plugin-standard to v4 ([5e48f2fa](https://github.com/lgaticaq/bip/commit/5e48f2fa810b3db79251e634c80ef50af463467b))
  *  update dependency nock to v9.6.1 ([bd39f6d5](https://github.com/lgaticaq/bip/commit/bd39f6d5a8970e9f617eaa824bd67150242b3418))
  *  update dependency eslint-plugin-node to v7 ([fea0d556](https://github.com/lgaticaq/bip/commit/fea0d556f2ecdd0321b45ab238b85f129290fd75))
  *  update dependency eslint-plugin-import to v2.14.0 ([7f1b6963](https://github.com/lgaticaq/bip/commit/7f1b69636b4b8e0219a76b69c1877c6d8c4278e8))
  *  update dependency lint-staged to v7.2.2 ([f5423fab](https://github.com/lgaticaq/bip/commit/f5423fab1c5b2ce5f8e868ae5758797765b7f5a8))
  *  update dependency nock to v9.6.0 ([7536bed0](https://github.com/lgaticaq/bip/commit/7536bed02baed009382eec7472ecfbbde7f93a43))
  *  update dependency nock to v9.5.0 ([9bbc115f](https://github.com/lgaticaq/bip/commit/9bbc115f337bf470befc71b07ccd56395e39800c))
  *  update dependency nock to v9.4.4 ([a9419615](https://github.com/lgaticaq/bip/commit/a9419615e3052110dc9ab089ac5c57342363a021))
  *  update dependency nock to v9.4.3 ([8c692091](https://github.com/lgaticaq/bip/commit/8c692091cff875e3c4bb9fd982ab88c9a074f260))
  *  update dependency nock to v9.4.2 ([b239cf02](https://github.com/lgaticaq/bip/commit/b239cf02597b2f68febbb513864589ea074a1653))
  *  update dependency nock to v9.4.1 ([167951a5](https://github.com/lgaticaq/bip/commit/167951a5231e569afecc161a0f2533f7a30713ed))
  *  update dependency nock to v9.4.0 ([0f62c88d](https://github.com/lgaticaq/bip/commit/0f62c88d506242902805ba30eac6bee1b0a35e74))
  *  update dependency eslint-plugin-import to v2.13.0 ([edc89404](https://github.com/lgaticaq/bip/commit/edc894048a8dc89199de750837de55c089d31e10))
  *  update dependency nock to v9.3.3 ([5d42db4a](https://github.com/lgaticaq/bip/commit/5d42db4a14ba98e96762b1192acff8645b25770c))
  *  update dependency lint-staged to v7.2.0 ([db730297](https://github.com/lgaticaq/bip/commit/db7302971878b704ee1fe6a2fa05a144f7d93342))
  *  update dependency nock to v9.3.2 ([f43d8353](https://github.com/lgaticaq/bip/commit/f43d835357372ec88c6cb6691058b9a6294b63e9))
  *  update dependency nock to v9.3.1 ([04e29acd](https://github.com/lgaticaq/bip/commit/04e29acdc5333fe601fadddeeb524607d073edf6))
  *  update dependency lint-staged to v7.1.3 ([ddba12ff](https://github.com/lgaticaq/bip/commit/ddba12ff7a605cf03e6e20f8d044233f7b5ea6a0))
  *  update dependency nock to v9.3.0 ([ccd7ac75](https://github.com/lgaticaq/bip/commit/ccd7ac75acd71bdef0e1162cf67815c28ecb00bd))
  *  update dependency eslint-plugin-promise to v3.8.0 ([bb3591ad](https://github.com/lgaticaq/bip/commit/bb3591ad7fef7b1118e5ab5292a528133e4eb352))
  *  update dependency lint-staged to v7.1.2 ([183a9549](https://github.com/lgaticaq/bip/commit/183a9549657f618737fa4c12445c38d49b513f29))
  *  update dependency mocha to v5.2.0 ([a70e7987](https://github.com/lgaticaq/bip/commit/a70e79874481cdbf964b5e49b355d449b3623cba))
  *  update dependency lint-staged to v7.1.1 ([177aa1df](https://github.com/lgaticaq/bip/commit/177aa1dfcad2f8f83f48a3889715cedfa0279277))
  *  update dependency eslint-plugin-import to v2.12.0 ([fdacef45](https://github.com/lgaticaq/bip/commit/fdacef457e3dcc0285a22d69ff52929a1559ff0d))
  *  update dependency nock to v9.2.6 ([042b2575](https://github.com/lgaticaq/bip/commit/042b25750d4076b52bc0794355f0f180befd2690))
  *  update dependency lint-staged to v7.1.0 ([e27a4af2](https://github.com/lgaticaq/bip/commit/e27a4af2d7f45514adc2b0533e44d2d9f1e94617))

##### Bug Fixes

* **parser:**  fix get negative balance ([4a18ebab](https://github.com/lgaticaq/bip/commit/4a18ebabf4075f1c2a6837d62ebc55a09228613e))

#### 3.0.1 (2018-05-03)

##### Chores

* **package:**  update package-lock.json ([51590c89](https://github.com/lgaticaq/bip/commit/51590c8938ccfbe575c91ff143b227e134f571d7))
* **deps:**
  *  update dependency lint-staged to v7.0.5 ([769ba105](https://github.com/lgaticaq/bip/commit/769ba10525ffa4ebec333b8d08f149354fbdcc36))
  *  update dependency eslint-plugin-standard to v3.1.0 ([8be51392](https://github.com/lgaticaq/bip/commit/8be51392c31a45296584c0825cc4802585602e6a))
  *  update dependency mocha to v5.1.1 ([bbb0b95c](https://github.com/lgaticaq/bip/commit/bbb0b95cf955d02e5ff24f628d220ffe94c36e8f))
  *  update dependency mocha to v5.1.0 ([3f4745f5](https://github.com/lgaticaq/bip/commit/3f4745f5da745d9c6c2890640c5c088d71e47eb1))
  *  update dependency eslint-plugin-import to v2.11.0 ([9113881d](https://github.com/lgaticaq/bip/commit/9113881da9e859546e7680ef70e73e12c1cf43de))
  *  update dependency nock to v9.2.5 ([af2c8f1d](https://github.com/lgaticaq/bip/commit/af2c8f1da0def6bbc54618e09aada5f4a01d3052))
  *  update dependency nock to v9.2.4 ([5081711e](https://github.com/lgaticaq/bip/commit/5081711e9cb3bdd99395ce540d7a383f4174b764))
  *  update dependency prettier-standard to v8.0.1 ([501d000f](https://github.com/lgaticaq/bip/commit/501d000fcffaa2ba2e860f35c61f19c48a7582d7))
  *  update dependency lint-staged to v7.0.4 ([12b9d515](https://github.com/lgaticaq/bip/commit/12b9d5155602f6b1f4aa6579864d53b6d87eb317))
  *  update dependency lint-staged to v7.0.3 ([b129da67](https://github.com/lgaticaq/bip/commit/b129da67eb314da912429177e2fd8a57456cfc2b))

##### Bug Fixes

* **parser:**  fix parse date ([b7b01357](https://github.com/lgaticaq/bip/commit/b7b0135716651e8abd70b977cd77012bef1a2c4f))

## 3.0.0 (2018-04-03)

##### Chores


test/test.js
'use strict'

const { describe, it, beforeEach } = require('mocha')
const { expect } = require('chai')
const path = require('path')
const nock = require('nock')

const lib = require('../src')

describe('bip', () => {
  describe('valid', () => {
    beforeEach(() => {
      nock.disableNetConnect()
      nock('http://pocae.tstgo.cl')
        .get('/PortalCAE-WAR-MODULE/')
        .reply(200, 'ok', {
          'set-cookie': [
            'JSESSIONID=517B19EEA6C1122EC934BDB624CFA413; Path=/PortalCAE-WAR-MODULE'
          ]
        })
      nock('http://pocae.tstgo.cl')
        .post('/PortalCAE-WAR-MODULE/SesionPortalServlet')
        .replyWithFile(200, path.join(__dirname, 'valid.html'))
    })

    it('should return a balance data', async () => {
      const data = await lib('11111111')
      expect(data.number).to.eql('11111111')
      expect(data.balance).to.eql(1180)
      expect(data.date).to.be.a('date')
      expect(data.message).to.eql('Tarjeta Valida')
      expect(data.valid).to.eql(true)
    })
  })

  describe('wrong status', () => {
    beforeEach(() => {
      nock.disableNetConnect()
      nock('http://pocae.tstgo.cl')
        .get('/PortalCAE-WAR-MODULE/')
        .reply(200, 'ok', {
          'set-cookie': [
            'JSESSIONID=517B19EEA6C1122EC934BDB624CFA413; Path=/PortalCAE-WAR-MODULE'
          ]
        })
      nock('http://pocae.tstgo.cl')
        .post('/PortalCAE-WAR-MODULE/SesionPortalServlet')
        .reply(200, '<table></table>')
    })

    it('should return an error', async () => {
      try {
        await lib('11111111')
      } catch (err) {
        expect(err.message).to.eql('Esta tarjeta no se puede cargar')
      }
    })
  })

  describe('wrong status code', () => {
    beforeEach(() => {
      nock.disableNetConnect()
      nock('http://pocae.tstgo.cl')
        .get('/PortalCAE-WAR-MODULE/')
        .reply(200, 'ok', {
          'set-cookie': [
            'JSESSIONID=517B19EEA6C1122EC934BDB624CFA413; Path=/PortalCAE-WAR-MODULE'
          ]
        })
      nock('http://pocae.tstgo.cl')
        .post('/PortalCAE-WAR-MODULE/SesionPortalServlet')
        .reply(301)
    })

    it('should return an error', async () => {
      try {
        await lib('11111111')
      } catch (err) {
        expect(err.message).to.eql('Request Failed. Status Code: 301')
      }
    })
  })

  describe('server error', () => {
    beforeEach(() => {
      nock.disableNetConnect()
      nock('http://pocae.tstgo.cl')
        .get('/PortalCAE-WAR-MODULE/')
        .reply(200, 'ok', {
          'set-cookie': [
            'JSESSIONID=517B19EEA6C1122EC934BDB624CFA413; Path=/PortalCAE-WAR-MODULE'
          ]
        })
      nock('http://pocae.tstgo.cl')
        .post('/PortalCAE-WAR-MODULE/SesionPortalServlet')
        .replyWithError('Server error')
    })

    it('should return an error', async () => {
      try {
        await lib('11111111')
      } catch (err) {
        expect(err.message).to.eql('Server error')
      }
    })
  })

  describe('wrong jssesionid', () => {
    beforeEach(() => {
      nock.disableNetConnect()
      nock('http://pocae.tstgo.cl')
        .get('/PortalCAE-WAR-MODULE/')
        .reply(200, 'ok')
    })

    it('should return an error', async () => {
      try {
        await lib('11111111')
      } catch (err) {
        expect(err.message).to.eql('Error al obtener el balance')
      }
    })
  })

  describe('wrong status code in jssesionid', () => {
    beforeEach(() => {
      nock.disableNetConnect()
      nock('http://pocae.tstgo.cl')
        .get('/PortalCAE-WAR-MODULE/')
        .reply(301)
    })

    it('should return an error in jssesionid', async () => {
      try {
        await lib('11111111')
      } catch (err) {
        expect(err.message).to.eql('Request Failed. Status Code: 301')
      }
    })
  })

  describe('server error', () => {
    beforeEach(() => {
      nock.disableNetConnect()
      nock('http://pocae.tstgo.cl')
        .get('/PortalCAE-WAR-MODULE/')
        .replyWithError('Server error')
    })

    it('should return an error', async () => {
      try {
        await lib('11111111')
      } catch (err) {
        expect(err.message).to.eql('Server error')
      }
    })
  })
})


src/index.js
'use strict'

const http = require('http')
const querystring = require('querystring')
const cheerio = require('cheerio')

/**
 * The built http.Agent object.
 * @external "http.Agent"
 * @see https://nodejs.org/api/http.html#http_class_http_agent
 */

/**
 * Get JSESSIONID from response headers
 * @param external:"http.Agent" agent
 * @returns {Promise<String>} JSESSIONID
 */
const getSessionId = agent => {
  return new Promise((resolve, reject) => {
    const options = {
      hostname: 'pocae.tstgo.cl',
      port: 80,
      path: '/PortalCAE-WAR-MODULE/',
      method: 'GET',
      agent
    }
    const req = http.request(options, res => {
      const { statusCode } = res
      let err
      if (statusCode !== 200) {
        err = new Error(`Request Failed. Status Code: ${statusCode}`)
      }
      if (err) {
        res.resume()
        return reject(err)
      }
      res.resume()
      try {
        resolve(res.headers['set-cookie'][0].split('; ')[0].split('=')[1])
      } catch (err) {
        reject(new Error('Error al obtener el balance'))
      }
    })
    req.on('error', err => reject(err))
    req.end()
  })
}

/**
 * The built http.Agent object.
 * @external "http.Agent"
 * @see https://nodejs.org/api/http.html#http_class_http_agent
 */

/**
 * Get balance of bip card
 * @param external:"http.Agent" agent
 * @param {String|Number} number Bip number card
 * @param {String} sessionId JSESSIONID
 * @returns {Promise<Object>} Object with number, message, balance and date
 */
const getBalance = (agent, number, sessionId) => {
  return new Promise((resolve, reject) => {
    const data = querystring.stringify({
      accion: 6,
      NumDistribuidor: 99,
      NomUsuario: 'usuInternet',
      NomHost: 'AFT',
      NomDominio: 'aft.cl',
      Trx: '',
      RutUsuario: 0,
      NumTarjeta: number,
      bloqueable: ''
    })
    const options = {
      hostname: 'pocae.tstgo.cl',
      port: 80,
      path: '/PortalCAE-WAR-MODULE/SesionPortalServlet',
      method: 'POST',
      headers: {
        Connection: 'keep-alive',
        Pragma: 'no-cache',
        'Cache-Control': 'no-cache',
        Origin: 'http://pocae.tstgo.cl',
        'Upgrade-Insecure-Requests': '1',
        'Content-Type': 'application/x-www-form-urlencoded',
        'User-Agent':
          'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/67.0.3381.0 Safari/537.36',
        Accept:
          'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8',
        Referer: 'http://pocae.tstgo.cl/PortalCAE-WAR-MODULE/',
        'Accept-Encoding': 'gzip, deflate',
        'Accept-Language': 'es-419,es;q=0.9',
        Cookie: `JSESSIONID=${sessionId}`
      },
      agent
    }
    const req = http.request(options, res => {
      if (res.statusCode !== 200) {
        reject(new Error(`Request Failed. Status Code: ${res.statusCode}`))
      } else {
        const rawData = []
        res.on('data', chunk => rawData.push(chunk))
        res.on('end', () => {
          try {
            const data = Buffer.concat(rawData).toString()
            const $ = cheerio.load(data)
            let [number, message, balance, date] = $('table')
              .find('td[bgcolor="#B9D2EC"]')
              .map((index, element) => $(element).text())
              .get()
            if (balance === '---' || typeof balance === 'undefined') {
              reject(new Error('Esta tarjeta no se puede cargar'))
            }
            let valid = false
            if (message === 'Contrato Activo') {
              message = 'Tarjeta Valida'
              valid = true
            }
            resolve({
              number,
              message,
              valid,
              balance: parseInt(balance.replace(/[$.]/g, ''), 10),
              date: new Date(
                date.replace(
                  /^(\d{2})\/(\d{2})\/(\d{4}) (\d{2}):(\d{2})$/,
                  '$3-$2-$1T$4:$5:00-03:00'
                )
              )
            })
          } catch (err) {
            reject(err)
          }
        })
      }
    })
    req.on('error', err => reject(err))
    req.write(data)
    req.end()
  })
}

/**
 * Get balance of bip card
 * @param {String|Number} number Bip number card
 * @returns {Promise<Object>} Object with number, message, balance and date
 */
module.exports = async number => {
  const agent = new http.Agent({ keepAlive: true })
  try {
    const sessionId = await getSessionId(agent)
    const balance = await getBalance(agent, number, sessionId)
    return balance
  } catch (err) {
    throw err
  }
}
